scylla/src/transport/downgrading_consistency_retry_policy.rs

```diff
@@ -165,3 +165,466 @@ impl RetrySession for DowngradingConsistencyRetrySession {
+                    true,
+                    cl
+                )),
+                RetryDecision::RetrySameNode(cl)
+            );
+            assert_eq!(
+                policy.decide_should_retry(make_query_info_with_cl(
+                    &enough_responses_no_data,
+                    true,
+                    cl
+                )),
+                RetryDecision::DontRetry
+            );
+        }
+        // Enough responses but data_present == true - coordinator probably timed out
+        // waiting for read-repair acknowledgement.
+        let enough_responses_with_data = QueryError::DbError(
+            DbError::ReadTimeout {
+                consistency: LegacyConsistency::Regular(Consistency::Two),
+                received: 2,
+                required: 2,
+                data_present: true,
+            },
+            String::new(),
+        );
+
+        for &cl in CONSISTENCY_LEVELS {
+            // Not idempotent
+            let mut policy = DowngradingConsistencyRetryPolicy::new().new_session();
+            assert_eq!(
+                policy.decide_should_retry(make_query_info_with_cl(
+                    &enough_responses_with_data,
+                    false,
+                    cl
+                )),
+                RetryDecision::DontRetry
+            );
+
+            // Idempotent
+            let mut policy = DowngradingConsistencyRetryPolicy::new().new_session();
+            assert_eq!(
+                policy.decide_should_retry(make_query_info_with_cl(
+                    &enough_responses_with_data,
+                    true,
+                    cl
+                )),
+                RetryDecision::DontRetry
+            );
+        }
+
+        // Not enough responses, data_present == true
+        let received = 1;
+        let not_enough_responses_with_data = QueryError::DbError(
+            DbError::ReadTimeout {
+                consistency: LegacyConsistency::Regular(Consistency::Two),
+                received,
+                required: 2,
+                data_present: true,
+            },
+            String::new(),
+        );
+        for &cl in CONSISTENCY_LEVELS {
+            // Not idempotent
+            let mut policy = DowngradingConsistencyRetryPolicy::new().new_session();
+            assert_eq!(
+                policy.decide_should_retry(make_query_info_with_cl(
+                    &not_enough_responses_with_data,
+                    false,
+                    cl
+                )),
+                max_likely_to_work_cl(received, cl)
+            );
+
+            // Idempotent
+            let mut policy = DowngradingConsistencyRetryPolicy::new().new_session();
+            assert_eq!(
+                policy.decide_should_retry(make_query_info_with_cl(
+                    &not_enough_responses_with_data,
+                    true,
+                    cl
+                )),
+                max_likely_to_work_cl(received, cl)
+            );
```

Comment:
Would be nice to check that, afterwards, the policy returns DontRetry. For better realism, don't forget to adjust CL for the second attempt. Please do it for both non-idempotent and idempotent case.